### PR TITLE
Add credit transfer history records

### LINKED ISSUE #74: Track credit transfer history for underlying credits
On the `polygon-digital-carbon` subgraph, add entities and fields needed to trace back the transfer history from on chain origination to final retirement.

- Record movements based on FIFO for pool holdings
- Show registry serial numbers if applicable on bridging

### PR DESCRIPTION
**Pull request by cujowolf**
This addition to the `polygon-digital-carbon` subgraph has the goal to create records that can be used to trace the transfer history of a specific carbon credit given a transaction hash back to the original bridging or issuance event on chain. 

This has been prepared using a first in, first out (FIFO) method when interacting with carbon pools.

Preview endpoint: https://api.thegraph.com/subgraphs/id/QmbVcXWp13DAqZGb6pLiXjFYr4QVUFtXeWEt22xq6BAkVm

Resolves #74 #76 

## Schema changes
New enum `ProvenanceType` with values:
- `ORIGINATION` - Initial bridging or issuance of a credit on-chain.
- `TRANSFER` - A standard transfer between two accounts.
- `RETIREMENT` - Final consumption and retirement of the credit.

New entity `ProvenanceRecord`
- `id` - Byte value from concatenation of token address, holding address, and incrementing record count on the credit.
- `token` - Credit token address.
- `transactionHash` - Transaction hash for record creation.
- `registrySerialNumbers` - Array of any serial numbers that are associated with this record. Only applicable to `ORIGINATION` record types.
- `priorRecords` - Array of cumulative prior `ProvenanceRecord`s linked to the current record.
- `sender` - Sender of the credits.
- `receiver` - Receiver of the credits.
- `originalAmount` - Initial transfer amount. Does not get updated post entity creation.
- `remainingAmount` - Amount left to transfer on this record. Used to facilitate FIFO calculations.
- `createdAt` - Timestamp of record creation.
- `updatedAt` - Timestamp of record update.

New fields for entity `CarbonCredit`:
- `provenanceCount` - Total provenance record count associated with this credit.
- `lastBatchId` - Last Toucan batch NFT ID that was fractionalized. This is used to pull in serial numbers for `ORIGINATION` records.

New entity `ToucanBatch`:
- `id` - Token ID of the batch.
- `registrySerialNumbers` - Array of serial numbers associated with this batch ID. Set from the `BatchUpdated` event.
- `creationTransactionHash` - Transaction hash for creation that is recorded in the Verra registry.

New fields for entity `Holding`:
- `activeProvenanceRecords` - Array of `ProvenanceRecord`s that are still being held by the account.
- `historicalProvenanceRecords` - Array of all `ProvenanceRecord`s that have been held by the account.

## Processing Logic Changes
Previously if an accounts holding balance became zero, the associated holding entity was removed from the store. With the need to track historical records, the entity is now removed from the store if the token balance is zero and there are no associated `ProvenanceRecord`s with this holding.

### CODE REVIEW

**Comment by cujowolf** [comment]
Example query showing all records associated with a specific transaction:
`query CreditTransaction {
  provenanceRecords(
    where: {transactionHash: "0x367c613ba894eac64ad045eb7f67ea5df552c3250b07b0ac9048bbff68c2c005"}
  ) {
    transactionType
    sender
    receiver
    originalAmount
    remainingAmount
    registrySerialNumbers
    transactionHash
    priorRecords(orderBy: createdAt, orderDirection: desc) {
      transactionType
      sender
      receiver
      originalAmount
      remainingAmount
      registrySerialNumbers
    }
  }
}`

**Comment by cujowolf** [comment]
Updated preview link with new Qm hash.

Add changes to resolve #76

**Review by 0xAurelius — APPROVED** [review]
Overall looks good, one minor question

File: polygon-digital-carbon/schema.graphql [hunk 1]
```diff
@@ -130,8 +174,11 @@ type CarbonCredit @entity {
   "All retirement events for this credit"
   retires: [Retire!]! @derivedFrom(field: "credit")
 
-  "Current pool balances for this credit"
-  poolBalances: [CarbonPoolCreditBalance!]! @derivedFrom(field: "credit")
```
[the remark this answers is not in the document]
Comment (cujowolf) [reply polygon-digital-carbon/schema.graphql:134]: Good catch, this should not have been removed. It's meant as just a reference entry so you can see what the pool balances are if you are already traversing the credit entity.

**Pull request merged**